[Mailer] Extract transport factory and allow create custom transports

Koc opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes, failure unrelated (master hasn't this PR)
| Fixed tickets | #31385, #32523
| License       | MIT
| Doc PR        | TBD

Alternative approach to allow create custom transports and register DSN for it. Replaces #31931, #31935 . Similar to already existent TansportFactory from Messenger.

TODO:
 - [x] Update changelog
 - [x] Add more tests for factories
 - [x] Add test for configuration + DI extension

Koc committed:
Add transport factories (closes #31385, closes #32523)

nicolas-grekas commented on src/Symfony/Component/Mailer/Transport/AbstractTransportFactory.php:
```diff
@@ -0,0 +1,24 @@
+namespace Symfony\Component\Mailer\Transport;
+
+use Psr\Log\LoggerInterface;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
```
if you can use the one from Contracts its better

Koc replied:
I've created separate PR for that: #31956

ajgarlag commented on src/Symfony/Component/Mailer/Exception/InvalidDsnException.php:
```diff
@@ -0,0 +1,52 @@
+            ],
+            'mandrill' => [
+                'class' => Bridge\Mailchimp\Factory\MandrillTransportFactory::class,
+                'package' => 'symfony/amazon-mailer',
```
```suggestion
                'package' => 'symfony/mailchimp-mailer',
```

onEXHovia commented on src/Symfony/Component/Mailer/Transport/Dsn.php:
```diff
@@ -0,0 +1,94 @@
+    private $options;
+
+    public function __construct(string $scheme, string $host, ?int $port, string $user, string $pass, array $options = [])
+    {
```
Port argument optional to maybe move it to the end?

fabpot replied:
Looks more natural like this.

onEXHovia commented on src/Symfony/Component/Mailer/Transport/NullTransportFactory.php:
```diff
@@ -0,0 +1,34 @@
+ * @author Konstantin Myakshin <molodchick@gmail.com>
+ */
+class NullTransportFactory extends AbstractTransportFactory
+{
```
Mark all TransportFactory how `final`?

fabpot replied:
I agree, they should all be final.

fabpot commented on src/Symfony/Component/Mailer/Transport/Dsn.php:
```diff
@@ -0,0 +1,94 @@
+class Dsn
+{
+    private $scheme;
+
```
white line between private props can be removed

fabpot commented on src/Symfony/Component/Mailer/Exception/InvalidDsnException.php:
```diff
@@ -0,0 +1,64 @@
+        return new self(sprintf('The "%s" scheme is not supported for mailer "%s".', $dsn->getScheme(), $dsn->getHost()));
+    }
+
+    public static function unsupportedHost(Dsn $dsn): self
```
We are using the Doctrine's style of doing exceptions. You need to create 2 different exception classes instead.

fabpot reviewed (changes requested):
Looks promising. Can you finish the PR? We should probably remove the current way or at least deprecate it.

Koc commented:
@fabpot I've fixed comments from code review. Now working on adding tests (I've updated TODO list in PR header)

fabpot commented on src/Symfony/Component/Mailer/CHANGELOG.md:
```diff
@@ -7,6 +7,8 @@ CHANGELOG
  * [BC BREAK] Transports depend on `Symfony\Contracts\EventDispatcher\EventDispatcherInterface`
    instead of `Symfony\Component\EventDispatcher\EventDispatcherInterface`.
 
+ * [BC BREAK] PLACEHOLDER.
```
Don't forget to change it to something explicit :)

fabpot commented on src/Symfony/Component/Mailer/TransportFactory.php:
```diff
@@ -0,0 +1,75 @@
+ * @author Fabien Potencier <fabien@symfony.com>
+ * @author Konstantin Myakshin <molodchick@gmail.com>
+ */
+class TransportFactory
```
What about keeping the `Transport` class name (I don't like classes with a name that end with `Factory`)? And if `$factories` is empty, fall back to the built-in transports. That way, we keep the current simple way of creating a Transport (without using Symfony full-stack) and we keep BC as well.

kevin-verschaeve commented on src/Symfony/Component/Mailer/Transport/Dsn.php:
```diff
@@ -0,0 +1,93 @@
+
+    public function getOption(string $key, $default = null)
+    {
+        return \array_key_exists($key, $this->options) ? $this->options[$key] : $default;
```
```suggestion
        return $this->options[$key] ?? $default;
```
?

Taluu replied:
If the value is `null` and the default isn't, it should be returned imo

Taluu commented on src/Symfony/Component/Mailer/Transport/Dsn.php:
```diff
@@ -0,0 +1,93 @@
+
+        $user = urldecode($parsedDsn['user'] ?? '');
+        $pass = urldecode($parsedDsn['pass'] ?? '');
+        parse_str($parsedDsn['query'] ?? '', $query);
```
wouldn't it be better to have `null` values for `user`, `password` and so on instead of empty string if not set ?

Taluu reviewed (approved):
Awesome. :+1:

Koc commented:
I've fixed tests and removed BC-breaks, so bc-break label non-actual anymore. More tests for all factories should be added, but I need 1 week more due to vacation.

BTW PR is ready for another round of review.

fabpot commented on src/Symfony/Component/Mailer/Transport/Smtp/EsmtpTransportFactory.php:
```diff
@@ -0,0 +1,49 @@
+            $dsn->getOption('auth_mode'),
+            $this->dispatcher,
+            $this->logger
+        );
```
Could be on one line

fabpot reviewed (approved):
Looks good to me. Thanks for the hard work so far!

Koc commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/mailer.xml:
```diff
@@ -12,12 +12,13 @@
         <service id="mailer" alias="mailer.mailer" />
         <service id="Symfony\Component\Mailer\MailerInterface" alias="mailer.mailer" />
 
+        <service id="mailer.transport_factory" class="Symfony\Component\Mailer\TransportFactory">
```
Should be `Transport`

Koc commented on src/Symfony/Component/Mailer/Transport/Dsn.php:
```diff
@@ -0,0 +1,94 @@
+        $this->host = $host;
+        $this->port = $port;
+        $this->user = $user;
+        $this->password = $password;
```
Need to reorder as constructor arguments

fabpot commented:
Not sure if this PR is ready now but fabbot has an interesting patch to apply.

Koc commented:
Almost ready. Failure caused because #32553 not merged yet.

fabpot commented:
@Koc #32553 has been merged now

Koc commented:
Finally PR is ready 🎉 

Travis is green except `deps=high` (because master hasn't this PR yet). To be honest I still haven't performed manual testing of the branch, but we already have test for DI extension and each factory properly tested.

fabpot commented:
Thank you @Koc.

stof commented on src/Symfony/Component/Mailer/Transport/Dsn.php:
```diff
@@ -0,0 +1,89 @@
+        }
+
+        $user = urldecode($parsedDsn['user'] ?? null);
+        $password = urldecode($parsedDsn['pass'] ?? null);
```
this is broken. It is passing `null` to `urldecode`, and it won't get `null` as `$user` then.

fabpot replied:
see 288772478cb4d96d340bc946745a859a0b01334a

Koc replied:
But it already covered with test https://github.com/symfony/symfony/pull/31946/files#diff-e3a1e78245b60fed36ed6ab4e56023ceR48 . Strange. Maybe because `assertEquals` compares `''` equal to `null`

francoispluchino replied:
@Koc `assertEquals()` does not make a strict comparison unlike `assertSame()`. It is therefore recommended to use `assertSame()` (or `assertNull()` to check the null value).
